discussment-core/src/main/java/org/danekja/discussment/core/service/imp/DefaultPostReputationService.java

```diff
@@ -0,0 +1,73 @@
+package org.danekja.discussment.core.service.imp;
+
+import org.danekja.discussment.core.accesscontrol.service.AccessControlService;
+import org.danekja.discussment.core.accesscontrol.service.DiscussionUserService;
+import org.danekja.discussment.core.dao.PostReputationDao;
+import org.danekja.discussment.core.dao.UserPostReputationDao;
+import org.danekja.discussment.core.domain.Post;
+import org.danekja.discussment.core.domain.PostReputation;
+import org.danekja.discussment.core.domain.UserPostReputation;
+import org.danekja.discussment.core.service.PostReputationService;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Implementation of the PostReputationService interface.
+ *
+ * Date: 17.2.18
+ *
+ * @author Jiri Kryda
+ */
+public class DefaultPostReputationService implements PostReputationService {
+
+    private PostReputationDao postReputationDao;
+    private UserPostReputationDao userPostReputationDao;
+
+    private AccessControlService accessControlService;
+    private DiscussionUserService userService;
+
+    public DefaultPostReputationService(PostReputationDao postReputationDao,
+                                        UserPostReputationDao userPostReputationDao,
+                                        DiscussionUserService userService,
+                                        AccessControlService accessControlService){
+
+        this.postReputationDao = postReputationDao;
+        this.userPostReputationDao = userPostReputationDao;
+
+        this.userService = userService;
+        this.accessControlService = accessControlService;
+    }
+
+    public PostReputation createPostReputation (PostReputation entity){
+        return postReputationDao.save(entity);
+    }
+
+    public PostReputation getPostReputationByPost (Post post){
+        return postReputationDao.getByPost(post);
+    }
+
+    public void addLike(PostReputation postReputation){
```

Comment:
Both `addLike` and `addDislike` are naive implementations with the following issues:

* what if two users click like at the same time - it is possible for `post.likes` number to suddenly be out of sync with number of `UserReputation` instances.
* postReputation like attribute doesnt get updated in the postreputation object is in detached state.
* what if user has liked the post before and now clicks dislike? Suddenly he both liked and dislike the post at the same time
* multiple calls to either of the methods result in multiple UserPostReputation instances added
* if we need to increase likes anywhere else in code, we need to call `postReputation.setLikes(postReputation.getLikes() + 1)` everywhere. This is bad object design.

## Solutions (same order):

* use `synchronized` block on `postDao` (after changing `PostrReputation` to `@Embeddable` and updating the API to wrap all the calls in this method:
```
//this is not the exact resulting code, just demonstration of synchronized
synchronized (postDao) {
   post.addLike();
   userPostReputationDao.save(...);
}
```
* after updating the number of `likes` on `Post`, the `Post` needs to be saved using `postDao.save`
* we need to first check, if `UserPostReputation` exists for the `Post` and `User` and create new one only if there is none yet. If it exists, we need to update its value instead of creating a new one.
* fixed by the previous
* `PostReputation` needs to have public method for updating number of likes (calling setters is a bad practice leading to code duplication and difficult testability). Add methods like:
```
public void addLike() {
    this.likes++;
}
```
etc.